pipeline/contrib/external_plugins/models/base.py

```diff
@@ -0,0 +1,119 @@
+Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://opensource.org/licenses/MIT
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+"""
+
+from copy import deepcopy
+from abc import abstractmethod
+
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+
+from pipeline.contrib.external_plugins import exceptions
+from pipeline.contrib.external_plugins.models.fields import JSONTextField
+
+GIT = 'git'
+S3 = 's3'
+FILE_SYSTEM = 'fs'
+
+package_source_types = frozenset({
+    GIT,
+    S3,
+    FILE_SYSTEM
+})
+
+source_cls_factory = {
+
+}
+
+
+def package_source(cls):
+    source_cls_factory[cls.type()] = cls
+    return cls
+
+
+class SourceManager(models.Manager):
+
+    def create_source(self, name, packages, from_config, **kwargs):
+        create_kwargs = deepcopy(kwargs)
+        create_kwargs['name'] = name
+        create_kwargs['packages'] = packages
+        create_kwargs['from_config'] = from_config
+        return self.create(**create_kwargs)
+
+    def remove_source(self, source_id):
+        source = self.get(id=source_id)
+
+        if source.from_config:
+            raise exceptions.InvalidOperationException('Can not remove source create from config')
+
+        source.delete()
+
+    def update_source_from_config(self, configs):
+
+        sources_from_config = self.filter(from_config=True).all()
+        existing_source_names = {source.name for source in sources_from_config}
+        source_name_in_config = {config['name'] for config in configs}
+
+        invalid_source_names = existing_source_names - source_name_in_config
+
+        # remove invalid source
+        self.filter(name__in=invalid_source_names).delete()
+
+        # update and create source
+        for config in configs:
+            defaults = deepcopy(config['details'])
+            defaults['packages'] = config['packages']
+
+            self.update_or_create(
+                name=config['name'],
+                from_config=True,
+                defaults=defaults)
+
+
+class ExternalPackageSource(models.Model):
+    name = models.CharField(_(u'包源名'), max_length=128, unique=True)
+    from_config = models.BooleanField(_(u"是否是从配置文件中读取的"), default=False)
+    packages = JSONTextField(_(u"模块配置"))
+
+    objects = SourceManager()
+
+    class Meta:
+        abstract = True
+
+    @staticmethod
+    @abstractmethod
+    def type():
+        raise NotImplementedError()
+
+    @abstractmethod
+    def importer(self):
+        raise NotImplementedError()
+
+    @property
+    def modules(self):
+        modules = []
+
+        for _, package_info in self.packages.items():
```

Comment:
for package_info in self.packages.values():

Reply:
done